File: Networking/src/main/java/protocolBuffers/ProtocolBufferServicesProxy.java
package protocolBuffers;

import DTO.EventDTO;
import DTO.ParticipantDTO;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import contestUtils.IContestServices;
import contestUtils.IMainObserver;
import example.model.Event;
import example.model.Office;
import example.model.Participant;
import example.model.User;

import java.io.*;
import java.net.Socket;
import java.util.Collection;
import java.util.List;
import java.util.concurrent.BlockingDeque;
import java.util.concurrent.LinkedBlockingDeque;

public class ProtocolBufferServicesProxy implements IContestServices {
    private final String host;
    private final int port;
    private IMainObserver client;
    private InputStream input;
    private OutputStream output;
    private Socket connection;
    private final static Logger logger = LogManager.getLogger();

    private final BlockingDeque<SwimmingContestProtocol.SwimmingContestResponse> responseQueue;
    private volatile boolean finished;

    public ProtocolBufferServicesProxy(String host, int port) {
        this.host = host;
        this.port = port;
        responseQueue = new LinkedBlockingDeque<>();
    }

    @Override
    public User login(String userName, String password, IMainObserver client) throws Exception {
        initializeConnection();
        SwimmingContestProtocol.SwimmingContestRequest request = ProtoBuilderUtils.createLoginRequest(userName, password);
        sendRequest(request);
        SwimmingContestProtocol.SwimmingContestResponse response = readResponse();
        if (response.getType() == SwimmingContestProtocol.SwimmingContestResponse.Type.Ok) {
            this.client = client;
            return ProtoBuilderUtils.getUser(response);
        } else if (response.getType() == SwimmingContestProtocol.SwimmingContestResponse.Type.Error) {
            closeConnection();
            logger.error(response.getErrorMessage());
            throw new Exception(response.getErrorMessage());
        }
        return null;
    }

    @Override
    public void logout(User user, IMainObserver client) throws Exception {
        SwimmingContestProtocol.SwimmingContestRequest request = ProtoBuilderUtils.createLogoutRequest(user);
        sendRequest(request);
        SwimmingContestProtocol.SwimmingContestResponse response = readResponse();
        closeConnection();
        if (response.getType() == SwimmingContestProtocol.SwimmingContestResponse.Type.Error) {
            logger.error(response.getErrorMessage());
            throw new Exception(response.getErrorMessage());
        }
    }

    @Override
    public Collection<EventDTO> getEventsWithParticipantsCount() throws Exception {
        SwimmingContestProtocol.SwimmingContestRequest request = ProtoBuilderUtils.createGetEventsWithParticipantsCountRequest();
        sendRequest(request);
        SwimmingContestProtocol.SwimmingContestResponse response = readResponse();
        if (response.getType() == SwimmingContestProtocol.SwimmingContestResponse.Type.Error) {
            logger.error(response.getErrorMessage());
            throw new Exception(response.getErrorMessage());
        }
        return ProtoBuilderUtils.getEventDTOs(response);
    }

    @Override
    public Collection<ParticipantDTO> getParticipantsForEventWithCount(Long eventId) throws Exception {
        SwimmingContestProtocol.SwimmingContestRequest request = ProtoBuilderUtils.createGetParticipantsForEventWithCountRequest(eventId);
        sendRequest(request);
        SwimmingContestProtocol.SwimmingContestResponse response = readResponse();
        if (response.getType() == SwimmingContestProtocol.SwimmingContestResponse.Type.Error) {
            logger.error(response.getErrorMessage());
            throw new Exception(response.getErrorMessage());
        }
        return ProtoBuilderUtils.getParticipantDTOs(response);
    }

    @Override
    public void saveEventEntry(List<Office> offices) throws Exception {
        SwimmingContestProtocol.SwimmingContestRequest request = ProtoBuilderUtils.createCreateEventEntryRequest(offices);
        sendRequest(request);
    }

    @Override
    public Collection<Participant> findAllParticipants() throws Exception {
        SwimmingContestProtocol.SwimmingContestRequest request = ProtoBuilderUtils.createGetAllParticipantsRequest();
        sendRequest(request);
        SwimmingContestProtocol.SwimmingContestResponse response = readResponse();
        if (response.getType() == SwimmingContestProtocol.SwimmingContestResponse.Type.Error) {
            logger.error(response.getErrorMessage());
            throw new Exception(response.getErrorMessage());
        }
        return ProtoBuilderUtils.getParticipants(response);
    }

    @Override
    public Collection<Event> findAllEvents() throws Exception {
        SwimmingContestProtocol.SwimmingContestRequest request = ProtoBuilderUtils.createGetAllEventsRequest();
        sendRequest(request);
        SwimmingContestProtocol.SwimmingContestResponse response = readResponse();
        if (response.getType() == SwimmingContestProtocol.SwimmingContestResponse.Type.Error) {
            throw new Exception(response.getErrorMessage());
        }
        return ProtoBuilderUtils.getEvents(response);
    }

    @Override
    public void saveEvent(Event event) throws Exception {
        SwimmingContestProtocol.SwimmingContestRequest request = ProtoBuilderUtils.createCreateEventRequest(event);
        sendRequest(request);
    }

    @Override
    public void saveParticipant(Participant participant) throws Exception {
        SwimmingContestProtocol.SwimmingContestRequest request = ProtoBuilderUtils.createCreateParticipantRequest(participant);
        sendRequest(request);
    }

    private void closeConnection() {
        finished = true;
        try {
            input.close();
            output.close();
            connection.close();
            client = null;
        } catch (IOException e) {
            logger.error("Error while closing connection " + e.getMessage());
            e.printStackTrace();
        }
    }

    private SwimmingContestProtocol.SwimmingContestResponse readResponse() throws Exception {
        try {
            SwimmingContestProtocol.SwimmingContestResponse response = responseQueue.take();
            if (response.getType() == SwimmingContestProtocol.SwimmingContestResponse.Type.Error) {
                throw new Exception(response.getErrorMessage());
            }
            return response;
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
            throw new Exception("Interrupted while waiting for response");
        }
    }

    private void sendRequest(SwimmingContestProtocol.SwimmingContestRequest request) throws Exception {
        try {
            logger.info("Sending request..." + request);
            request.writeDelimitedTo(output);
            output.flush();
        } catch (IOException e) {
            logger.error("Sending request failed", e);
            throw new Exception("Sending request failed", e);
        }
    }

    private void initializeConnection() {
        try {
            connection = new Socket(host, port);
            output = connection.getOutputStream();
            output.flush();
            input = connection.getInputStream();
            finished = false;
            startResponseReader();
        } catch (IOException e) {
            logger.error("Error while initializing connection " + e.getMessage());
            e.printStackTrace();
        }
    }

    private void startResponseReader() {
        Thread responseReader = new Thread(new ReaderThread());
        responseReader.start();
    }

    private class ReaderThread implements Runnable {
        public void run() {
            while (!finished) {
                try {
                    SwimmingContestProtocol.SwimmingContestResponse response = SwimmingContestProtocol.SwimmingContestResponse.parseDelimitedFrom(input);
                    logger.info("response received " + response);
                    if (isUpdateResponse(response)) {
                        handleUpdate(response);
                    }
                    else {
                        try{
                            responseQueue.put(response);
                        }catch(InterruptedException e){
                            e.printStackTrace();
                        }
                    }
                } catch (Exception exception) {
                    logger.error("reading error: {}", exception);
                }
            }
        }
    }

    private boolean isUpdateResponse(SwimmingContestProtocol.SwimmingContestResponse response) {
        return response.getType() == SwimmingContestProtocol.SwimmingContestResponse.Type.NewParticipant
                || response.getType() == SwimmingContestProtocol.SwimmingContestResponse.Type.NewEvent
                || response.getType() == SwimmingContestProtocol.SwimmingContestResponse.Type.NewEventEntry
                || response.getType() == SwimmingContestProtocol.SwimmingContestResponse.Type.UpdatedEvents;
    }

    private void handleUpdate(SwimmingContestProtocol.SwimmingContestResponse response) throws Exception {
        switch (response.getType()) {
            case NewParticipant -> client.participantAdded(ProtoBuilderUtils.getParticipant(response));

            //case NewEvent -> client.eventAdded(ProtoBuilderUtils.getEvent(response));

            case UpdatedEvents -> client.newRegistration(ProtoBuilderUtils.getEventDTOs(response));
        }
    }
}

File: java-app/csharp-app/Networking/ProtocolBuffers/ProtocolBufferServicesProxy.cs
using System;
using System.Net.Sockets;
using Google.Protobuf;
using log4net;
using Org.Example.Protocolbuffers;
using Service;

using Participant = mpp_proiect_csharp_DianaGliga11.Model.Participant;
using User = mpp_proiect_csharp_DianaGliga11.Model.User;
using Event = mpp_proiect_csharp_DianaGliga11.Model.Event;
using EventDTO = mpp_proiect_csharp_DianaGliga11.Model.DTO.EventDTO;
using ParticipantDTO = mpp_proiect_csharp_DianaGliga11.Model.DTO.ParticipantDTO;
using Office = mpp_proiect_csharp_DianaGliga11.Model.Office;

namespace Networking.ProtocolBuffers
{
    public class ProtocolBufferServicesProxy : IContestServices
    {

        private readonly string _host;
        private readonly int _port;
        private IMainObserver _clientObserver;
        private TcpClient _connection;
        private NetworkStream _stream;
        private readonly Queue<SwimmingContestResponse> _responseQueue = new();
        private volatile bool _finished;

        private EventWaitHandle _waitHandle;

        private static readonly ILog Log = LogManager.GetLogger(typeof(ServicesProxy));

        public ProtocolBufferServicesProxy(string host, int port)
        {
            _host = host;
            _port = port;
            _waitHandle = new AutoResetEvent(false);
            _responseQueue = new Queue<SwimmingContestResponse>();
        }

        private void EnsureConnected()
        {
            if (_connection?.Connected == true) return;

            try
            {
                _connection = new TcpClient(_host, _port);
                _stream = _connection.GetStream();
                _finished = false;
                _waitHandle = new AutoResetEvent(false);
                StartReader();
                Log.Info($"Connected to {_host}:{_port}");
            }
            catch (Exception ex)
            {
                Log.Error($"Connection failed: {ex.Message}");
                throw;
            }
        }


        private void RunReader()
        {
            while (!_finished)
            {
                try
                {
                    SwimmingContestResponse response = SwimmingContestResponse.Parser.ParseDelimitedFrom(_stream);
                    Log.Info("Response received: " + response);
                    if (IsUpdateResponse(response))
                    {
                        HandleUpdate(response);
                    }
                    else
                    {
                        lock (_responseQueue)
                        {
                            _responseQueue.Enqueue(response);
                        }

                        _waitHandle.Set();
                    }
                }
                catch (Exception e)
                {
                    Log.Error("Error in Run Client");
                }
            }
        }

        private bool IsUpdateResponse(SwimmingContestResponse response)
        {
            return response.Type == SwimmingContestResponse.Types.Type.NewParticipant
                   || response.Type == SwimmingContestResponse.Types.Type.UpdatedEvents
                   || response.Type == SwimmingContestResponse.Types.Type.NewEvent;
        }

        private SwimmingContestResponse ReadResponse()
        {
            SwimmingContestResponse response = null;
            try
            {
                _waitHandle.WaitOne();
                lock (_responseQueue)
                {
                    response = _responseQueue.Dequeue();
                }
            }
            catch (Exception e)
            {
                Log.Error("Error while reading response");
            }

            return response;
        }

        private void SendRequest(SwimmingContestRequest req)
        {
            try
            {
                Log.Debug($"Sending {req.Type} request");
                req.WriteDelimitedTo(_stream);
                _stream.Flush();
            }
            catch (Exception ex)
            {
                Log.Error($"SendRequest failed: {ex.Message}");
                CloseConnection();
                throw;
            }
        }
        private void HandleUpdate(SwimmingContestResponse msg)
        {
            if (IsUpdateResponse(msg))
            {
                try
                {
                    switch (msg.Type)
                    {
                        case SwimmingContestResponse.Types.Type.NewParticipant:
                            _clientObserver.ParticipantAdded(ProtocolBuilderUtils.GetParticipant(msg));
                            break;
                        case SwimmingContestResponse.Types.Type.NewEvent:
                            break;
                        case SwimmingContestResponse.Types.Type.UpdatedEvents:
                            _clientObserver.EventEvntriesAdded(ProtocolBuilderUtils.GetEventDTOs(msg));
                            break;
                    }
                }
                catch (Exception e)
                {
                    Console.WriteLine(e.StackTrace);
                }
            }
        }
        
        public User Login(string username, string password, IMainObserver client)
        {
            try
            {
                EnsureConnected();
                _clientObserver = client;

                var request = new SwimmingContestRequest
                {
                    Type = SwimmingContestRequest.Types.Type.Login,
                    UserName = username,
                    Password = password
                };

                Log.Debug($"Sending login request for {username}");
                SendRequest(request);
            
                var response = ReadResponse();
                Log.Debug($"Received response type: {response?.Type}");

                if (response == null)
                    throw new Exception("Null response from server");

                if (response.Type == SwimmingContestResponse.Types.Type.Error)
                    throw new Exception(response.ErrorMessage);

                if (response.User == null)
                    throw new Exception("User data missing in response");

                return new User(response.User.UserName, response.User.Password) 
                { 
                    Id = response.User.Id 
                };
            }
            catch (Exception ex)
            {
                Log.Error($"Login failed: {ex.Message}");
                CloseConnection();
                throw;
            }
        }

        private void CloseConnection()
        {
            _finished = true;
            try
            {
                _stream.Close();
                _connection.Close();
                _waitHandle.Close();
            }
            catch (Exception e)
            {
                Log.Error("Error while closing connection");
            }
        }

        private void StartReader()
        {
            Thread worker = new Thread(RunReader);
            worker.Start();
        }

        public void Logout(User user, IMainObserver client)
        {
            SwimmingContestRequest request = ProtocolBuilderUtils.CreateLogoutRequest(user);
            SendRequest(request);

            SwimmingContestResponse response = ReadResponse();
            CloseConnection();

            if (response.Type == SwimmingContestResponse.Types.Type.Error)
            {
                Log.Error("Error while logging out");
            }
        }

        public List<EventDTO> GetEventsWithParticipantsCount()
        {
            SwimmingContestRequest request
                = ProtocolBuilderUtils.CreateGetEventsWithParticipantsCountRequest();
            SendRequest(request);

            SwimmingContestResponse response = ReadResponse();
            if (response.Type == SwimmingContestResponse.Types.Type.Error)
            {
                Log.Error("Error while getting events with participant count");
            }

            return ProtocolBuilderUtils.GetEventDTOs(response);
        }

        public List<ParticipantDTO> GetParticipantsForEventWithCount(long eventId)
        {
            SwimmingContestRequest request = ProtocolBuilderUtils.CreateGetParticipantsForEventWithCountRequest(eventId);
            SendRequest(request);
            SwimmingContestResponse response = ReadResponse();
            if (response.Type == SwimmingContestResponse.Types.Type.Error)
            {
                Log.Error("Error while getting events with participant count");
            }
            return ProtocolBuilderUtils.GetParticipantDTOs(response);
        }

        public List<Participant> GetAllParticipants()
        {
            SwimmingContestRequest request = ProtocolBuilderUtils.CreateGetAllParticipantsRequest();
            SendRequest(request);

            SwimmingContestResponse response = ReadResponse();
            if (response.Type == SwimmingContestResponse.Types.Type.Error)
            {
                Log.Error("Error while getting participants");
            }

            return ProtocolBuilderUtils.GetParticipants(response);
        }

        public List<Event> GetAllEvents()
        {
            SwimmingContestRequest request = ProtocolBuilderUtils.CreateGetAllEventsRequest();
            SendRequest(request);

            SwimmingContestResponse response = ReadResponse();
            if (response.Type == SwimmingContestResponse.Types.Type.Error)
            {
                Log.Error("Error while getting participants");
            }

            return ProtocolBuilderUtils.GetEvents(response);
        }

        public void SaveEventsEntries(List<Office> newEntries)
        {
            SwimmingContestRequest request = ProtocolBuilderUtils.CreateCreateEventEntryRequest(newEntries);
            SendRequest(request);
            SwimmingContestResponse response = ReadResponse();
            if (response.Type == SwimmingContestResponse.Types.Type.Error)
            {
                Log.Error("Error while saving events entries");
            }
        }

        public void SaveParticipant(Participant participant, IMainObserver sender)
        {
            try
            {
                var request = ProtocolBuilderUtils.CreateCreateParticipantRequest(participant);
                SendRequest(request);
                var response = ReadResponse();
            
                if (response.Type == SwimmingContestResponse.Types.Type.Error)
                    throw new Exception(response.ErrorMessage);
            }
            catch (Exception ex)
            {
                Log.Error($"SaveParticipant failed: {ex.Message}");
                throw;
            }
        }

    }
}